File: pm/src/main/java/org/sqlbuilder/pm/objects/PmPredicate.java
package org.sqlbuilder.pm.objects;

import org.sqlbuilder.common.HasId;
import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.common.NotNull;
import org.sqlbuilder.common.SetCollector;

import java.util.Comparator;
import java.util.Objects;

public class PmPredicate implements HasId, Insertable, Comparable<PmPredicate>
{
	public static final Comparator<PmPredicate> COMPARATOR = Comparator.comparing(PmPredicate::getPredicate);

	public static final SetCollector<PmPredicate> COLLECTOR = new SetCollector<>(COMPARATOR);

	public final String predicate;

	public static PmPredicate make(final String predicate)
	{
		var p = new PmPredicate(predicate);
		COLLECTOR.add(p);
		return p;
	}

	private PmPredicate(final String predicate)
	{
		this.predicate = predicate;
	}

	// A C C E S S

	public String getPredicate()
	{
		return predicate;
	}

	public String getWord()
	{
		return predicate.substring(0, predicate.indexOf('.'));
	}

	@Override
	public Integer getIntId()
	{
		return COLLECTOR.apply(this);
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		PmPredicate that = (PmPredicate) o;
		return predicate.equals(that.predicate);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(predicate);
	}

	// O R D E R

	@Override
	public int compareTo(@NotNull final PmPredicate that)
	{
		return COMPARATOR.compare(this, that);
	}

	// I N S E R T

	@Override
	public String dataRow()
	{
		return String.format("'%s'", predicate);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[PRED %s]", predicate);
	}
}

File: pm/src/main/java/org/sqlbuilder/pm/objects/PmPredicate.kt
package org.sqlbuilder.pm.objects

import org.sqlbuilder.common.HasId
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.SetCollector
import java.util.*

class PmPredicate private constructor(
    val predicate: String
) : HasId, Insertable, Comparable<PmPredicate> {

    val word: String
        get() = predicate.substring(0, predicate.indexOf('.'))

    override fun getIntId(): Int {
        return COLLECTOR.invoke(this)
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as PmPredicate
        return predicate == that.predicate
    }

    override fun hashCode(): Int {
        return Objects.hash(predicate)
    }

    // O R D E R

    override fun compareTo(that: PmPredicate): Int {
        return COMPARATOR.compare(this, that)
    }

    // I N S E R T

    override fun dataRow(): String {
        return "'$predicate'"
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[PRED $predicate]"
    }

    companion object {

        val COMPARATOR: Comparator<PmPredicate> = Comparator
            .comparing<PmPredicate, String> { it.predicate }

        val COLLECTOR = SetCollector<PmPredicate>(COMPARATOR)

        fun make(predicate: String): PmPredicate {
            val p = PmPredicate(predicate)
            COLLECTOR.add(p)
            return p
        }
    }
}
